lib/connection.ts

```diff
@@ -0,0 +1,355 @@
+  }
+
+  /**
+   * @property {string} id Returns the connection id.
+   * @readonly
+   */
+  get id(): string {
+    return this._connection.options.id!;
+  }
+
+  /**
+   * Creates a new amqp connection.
+   * @param {ConnectionOptions} [options] Options to be provided for establishing an amqp connection.
+   * @return {Promise<Connection>} Promise<Connection>
+   * - **Resolves** the promise with the Connection object when rhea emits the "connection_open" event.
+   * - **Rejects** the promise with an AmqpError when rhea emits the "connection_close" event while trying
+   * to establish an amqp connection.
+   */
+  open(): Promise<Connection> {
+    return new Promise((resolve, reject) => {
+      if (!this.isOpen()) {
+
+        let onOpen: Func<rhea.EventContext, void>;
+        let onClose: Func<rhea.EventContext, void>;
+        let waitTimer: any;
+
+        const removeListeners: Function = () => {
+          clearTimeout(waitTimer);
+          this._connection.removeListener(ConnectionEvents.connectionOpen, onOpen);
+          this._connection.removeListener(ConnectionEvents.connectionClose, onClose);
+          this._connection.removeListener(ConnectionEvents.disconnected, onClose);
+        };
+
+        onOpen = (context: rhea.EventContext) => {
+          removeListeners();
+          process.nextTick(() => {
+            log.connection("[%s] Resolving the promise with amqp connection.", this.id);
+            resolve(this);
+          });
+        };
+
+        onClose = (context: rhea.EventContext) => {
+          removeListeners();
+          const err = context.error || context.connection.error;
+          log.error("[%s] Error occurred while establishing amqp connection: %O",
+            this.id, err);
+          reject(err);
+        };
+
+        const actionAfterTimeout = () => {
+          removeListeners();
+          const msg: string = `Unable to open the amqp connection "${this.id}" due to operation timeout.`;
+          log.error("[%s] %s", this.id, msg);
+          reject(new Error(msg));
+        };
+
+        this._connection.once(ConnectionEvents.connectionOpen, onOpen);
+        this._connection.once(ConnectionEvents.connectionClose, onClose);
+        this._connection.once(ConnectionEvents.disconnected, onClose);
+        waitTimer = setTimeout(actionAfterTimeout, defaultOperationTimeoutInSeconds * 1000);
+        this._connection.connect();
+      } else {
+        resolve(this);
+      }
+    });
+  }
+
+
+  /**
+   * Closes the amqp connection.
+   * @return {Promise<void>} Promise<void>
+   * - **Resolves** the promise when rhea emits the "connection_close" event.
+   * - **Rejects** the promise with an AmqpError when rhea emits the "connection_error" event while trying
+   * to close an amqp connection.
+   */
+  close(): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      log.error("[%s] The connection is open ? -> %s", this.id, this.isOpen());
+      if (this.isOpen()) {
```

Comment:
what does this return if you are currently in the process of opening the connection? i'm assuming false, in which case you could end up with an open connection if you call `open()` then `close()` before the open is done